src/mapping/nav2_map_server/src/map_reps/occupancy_grid.cpp

```diff
@@ -0,0 +1,353 @@
+// added to provide support for code written under the yaml-cpp 0.3 API.
+
+template <typename T>
+void operator>>(const YAML::Node &node, T &i)
+{
+	i = node.as<T>();
+}
+
+#endif
+
+// Interface //
+
+void OccGridLoader::loadMapInfoFromFile(std::string fname)
+{
+	std::ifstream fin(fname.c_str());
+	if (fin.fail())
+	{
+		fprintf(stderr, "[ERROR] [map_server] Map_server could not open %s\n", fname.c_str());
+		exit(-1);
+	}
+
+#ifdef HAVE_YAMLCPP_GT_0_5_0
+	// The document loading process changed in yaml-cpp 0.5.
+	//YAML::Node doc = YAML::Load(fin);
+	YAML::Node doc = YAML::LoadFile(fname);
+#else
+	YAML::Parser parser(fin);
+	YAML::Node doc;
+	parser.GetNextDocument(doc);
+#endif
+	try
+	{
+		doc["resolution"] >> res;
+	}
+	catch (YAML::InvalidScalar)
+	{
+		fprintf(stderr, "[ERROR] [map_server]: The map does not contain a resolution tag or it is invalid.\n");
+
+		exit(-1);
+	}
+
+	try
+	{
+		doc["negate"] >> negate;
+	}
+	catch (YAML::InvalidScalar)
+	{
+		fprintf(stderr, "[ERROR] [map_server]: The map does not contain a negate tag or it is invalid.\n");
+
+		exit(-1);
+	}
+	try
+	{
+		doc["occupied_thresh"] >> occ_th;
+	}
+	catch (YAML::InvalidScalar)
+	{
+		fprintf(stderr, "[ERROR] [map_server]: The map does not contain an occupied_thresh tag or it is invalid.\n");
+
+		exit(-1);
+	}
+	try
+	{
+		doc["free_thresh"] >> free_th;
+	}
+	catch (YAML::InvalidScalar)
+	{
+		fprintf(stderr, "[ERROR] [map_server]: The map does not contain a free_thresh tag or it is invalid.\n");
+
+		exit(-1);
+	}
+	try
+	{
+		std::string modeS = "";
+		doc["mode"] >> modeS;
+
+		if (modeS == "trinary")
+			mode = TRINARY;
+		else if (modeS == "scale")
+			mode = SCALE;
+		else if (modeS == "raw")
+			mode = RAW;
+		else
+		{
+			fprintf(stderr, "Invalid mode tag \"%s\".\n", modeS.c_str());
+
+			exit(-1);
+		}
+	}
+	catch (YAML::Exception)
+	{
+		//fprintf(stdout,"[DEBUG] [map_server] The map does not contain a mode tag or it is invalid... assuming Trinary\n");
+		mode = TRINARY;
+	}
+	try
+	{
+		doc["origin"][0] >> origin[0];
+		doc["origin"][1] >> origin[1];
+		doc["origin"][2] >> origin[2];
+	}
+	catch (YAML::InvalidScalar)
+	{
+		fprintf(stderr, "[ERROR] [map_server]: The map does not contain an origin tag or it is invalid.\n");
+
+		exit(-1);
+	}
+	try
+	{
+		doc["image"] >> mapfname;
+		// TODO: make this path-handling more robust
+		if (mapfname.size() == 0)
+		{
+			fprintf(stderr, "[ERROR] [map_server]: The image tag cannot be an empty string.\n");
+
+			exit(-1);
+		}
+		if (mapfname[0] != '/')
+		{
+			// dirname can modify what you pass it
+			char *fname_copy = strdup(fname.c_str());
+			mapfname = std::string(dirname(fname_copy)) + '/' + mapfname;
+			free(fname_copy);
+		}
+	}
+	catch (YAML::InvalidScalar)
+	{
+		fprintf(stderr, "[ERROR] [map_server]: The map does not contain an image tag or it is invalid.\n");
+
+		exit(-1);
+	}
+}
+
+void OccGridLoader::loadMapFromFile(std::string mapfname)
```

Comment:
Should this function be instead on the utils folder?
On the planner_tester, I also have code for reading an image file and generating an occupancy grid msg.

```C++
nav_msgs::msg::OccupancyGrid loadMapFromFile(
  const std::string image_file_name, const double resolution, const bool negate,
  const double occupancy_threshold, const double free_threshold, const geometry_msgs::msg::Twist origin,
  const MapMode mode)
```

Reply:
Yes, we should be on the lookout for utility code to factor it out into a utility library.